agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+        if (message != null)
+            synchronized(receivedMessages) { receivedMessages.add(message) }
+    }
+
+    override fun onError(ex: Exception?) {
+        log.warn(ex) { "AgentConnection error" }
+    }
+
+    fun sendJson(obj: Any) = send(SerializationUtils.anyToJSON(obj))
+
+    fun popMessageOfType(type: String): String? {
+        synchronized(receivedMessages) {
+            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
+            if (result != null)
+                receivedMessages.remove(result)
+
+            return result
+        }
+    }
+
+    fun waitForMessageOfType(type: String): String {
+        var messageOfType: String? = null
+        while (messageOfType == null) {
+            sleep(500)
+            messageOfType = popMessageOfType(type)
+        }
+        return messageOfType
+    }
+
+    inline fun <reified T : Any> waitForMessageOfType(type: String): T {
+        return SerializationUtils.jSONToAny(waitForMessageOfType(type))
+    }
+
+    fun sendTypedMessage(message: TypedBodyMessage, counterParty: IndyParty) = sendJson(SendMessage(counterParty.did, message))
+    inline fun <reified T : Any> sendTypedMessage(message: T, counterParty: IndyParty) = sendTypedMessage(TypedBodyMessage(message, T::class.java.canonicalName), counterParty)
+    inline fun <reified T : Any> popTypedMessage(): T? {
+        synchronized(receivedMessages) {
+            val message = receivedMessages
+                    .filter { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(MESSAGE_TYPES.MESSAGE_RECEIVED) }
+                    .find { SerializationUtils.jSONToAny<MessageReceived>(it).message.content.clazz == T::class.java.canonicalName }
+
+            if (message != null) {
+                val result = SerializationUtils.jSONToAny<T>(
+                        SerializationUtils.anyToJSON(
+                                SerializationUtils.jSONToAny<MessageReceived>(message).message.content.message
+                        )
+                )
+                receivedMessages.remove(message)
+                return result
+            }
+            return null
+        }
+    }
+
+    inline fun <reified T : Any> waitForTypedMessage(): T {
+        var messageOfType: T? = null
+        while (messageOfType == null) {
+            sleep(500)
+            messageOfType = popTypedMessage()
+        }
+        return messageOfType
+    }
+}
+
+object MESSAGE_TYPES {
+    val CONN_BASE = "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/"
+    val ADMIN_CONNECTIONS_BASE = "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/admin_connections/1.0/"
+    val ADMIN_BASE = "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/admin/1.0/"
+    val ADMIN_WALLETCONNECTION_BASE = "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/admin_walletconnection/1.0/"
+    val ADMIN_BASICMESSAGE_BASE = "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/admin_basicmessage/1.0/"
```

Comment:
Well, there is very rely on PythonRefAgent. DIDs for other agent will be different..

Reply:
This message types are from API requirements, we cant change them atm

Reply:
If we migrate on another cloud agent there are another ids, I suppose to have it configurable. I think the current connector is gonna be WSDriver and able to connect to any agent who rely on Indy's specification. Am I right? In the body of messages should be configurable

Reply:
For PythonRefAgent its hard coded, for VCX there is no agent DID in this messages.
I would suggest to wait for spec to stabilize.

Reply:
ok